Remotion/Data/DomainObjects/Mapping/TypeDefinitionWalker1.cs

```diff
@@ -0,0 +1,86 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Linq;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Mapping
+{
+  /// <summary>
+  /// Implements a <see cref="ITypeDefinitionVisitor{T}"/> that visits the type definitions in a depth-first manner
+  /// aborting the walk and returning the first value of <typeparamref name="T"/> that fulfills a specified predicate.
+  /// </summary>
+  public abstract class TypeDefinitionWalker<T> : ITypeDefinitionVisitor<T>
+  {
+    protected readonly TypeDefinitionWalkerDirection Direction;
+    protected readonly Func<T?, bool> CompletionPredicate;
+
+    protected TypeDefinitionWalker (TypeDefinitionWalkerDirection direction, Func<T?, bool> completionPredicate)
```

Comment:
ah, oh, yeah. I thought we do have a transparent conversion. and the last time I used this, I was using linq syntax isntead of extension methods and the c# compiler invokes the delegate instead of passing it on to avoid this conflict. okay.